### PR TITLE
imgcodecs: implement imencodemulti()

### LINKED ISSUE #26207: cv::imencode function fails writing vector<Mat>
### Describe the feature and motivation

test code :
```
    vector<Mat> imgs;
    Mat img(100, 100, CV_8UC1);
    imgs.push_back(img);
    std::vector<uchar> buf;
    bool ret_image = imencode(".tiff", imgs, buf);
    bool ret_multipage = imencode(".tiff", imgs, buf);
    printf("%d\n", ret_image);
    printf("%d\n", ret_multipage);
    imgs.push_back(img.clone());
    ret_multipage = imencode(".tiff", imgs, buf);
    printf("%d\n", ret_multipage);
    return 0;
```

### Additional context

output of test code:
```
1
1
OpenCV TIFF: TIFFLinkDirectory: Warning, Error fetching directory count.
[ WARN:0@0.006] global grfmt_tiff.cpp:1378 cv::TiffEncoder::writeLibTiff OpenCV TIFF(line 1378): failed TIFFWriteDirectory(tif)
OpenCV: terminate handler is called! The last OpenCV error is:
OpenCV(4.10.0-dev) Error: Unspecified error (OpenCV TIFF: failed TIFFWriteDirectory(tif)) in cv::TiffEncoder::writeLibTiff, file D:\GitHub\opencv\modules\imgcodecs\src\grfmt_tiff.cpp, line 1378
```

### PR DESCRIPTION
**Pull request by Kumataro**
Close #26207 
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/imgcodecs/test/test_read_write.cpp [hunk 1]
```diff
@@ -520,6 +520,22 @@ TEST(ImgCodecs, multipage_collection_two_iterator_operatorpp)
          EXPECT_TRUE(cv::norm(img1, img[i], NORM_INF) == 0);
     }
 }
+
+// See https://github.com/opencv/opencv/issues/26207
+TEST(Imgcodecs, imencodemulti_regression_26207)
+{
+    vector<Mat> imgs;
+    Mat img(100, 100, CV_8UC1);
+    imgs.push_back(img);
```
Comment (sturkmen72) [inline modules/imgcodecs/test/test_read_write.cpp:529]: @Kumataro could you test 
```
   vector<Mat> imgs;
    Mat img(100, 100, CV_8UC1);
    imgs.push_back(img);
    imgs.push_back(img.clone());
```

Comment (Kumataro) [reply modules/imgcodecs/test/test_read_write.cpp:529]: Thank you for your comment ! I mistakenly thought that it was enough to support multiple images for imencode().

But it seems that encoding multiple images requests reading callback function. After appending it, it works well.

Comment (sturkmen72) [reply modules/imgcodecs/test/test_read_write.cpp:529]: > Thank you for your comment ! I mistakenly thought that it was enough to support multiple images for imencode().
> 
> But it seems that encoding multiple images requests reading callback function. After appending it, it works well.

Thank you for your contribution. to me looks perfect.

File: modules/imgcodecs/src/loadsave.cpp [hunk 2]
```diff
@@ -1121,27 +1121,43 @@ bool imdecodemulti(InputArray _buf, int flags, CV_OUT std::vector<Mat>& mats, co
...
+
+        const int channels = image.channels();
+        CV_Assert( channels == 1 || channels == 3 || channels == 4 );
+
         Mat temp;
-        image.convertTo(temp, CV_8U);
-        image = temp;
+        if( !encoder->isFormatSupported(image.depth()) )
+        {
+            CV_Assert( encoder->isFormatSupported(CV_8U) );
```
Comment (sturkmen72) [inline modules/imgcodecs/src/loadsave.cpp:1155]: i think all encoders supports CV_8U so imho no need to this check

Comment (Kumataro) [reply modules/imgcodecs/src/loadsave.cpp:1155]: > i think all encoders supports CV_8U so imho no need to this check

Currently almost encoders supports CV_8U, but OpenEXR supports only CV_32F. And I think keeping behaviour between `imwrite()` and `imencode()` makes easy to debug encoders for current and future.

https://docs.opencv.org/4.10.0/d4/da8/group__imgcodecs.html#ga8ac397bd09e48851665edbe12aa28f25

> With OpenEXR encoder, only 32-bit float (CV_32F) images can be saved.
> 8-bit unsigned (CV_8U) images are not supported.

https://github.com/opencv/opencv/blob/450e741f8d53ff12b4e194c7762adaefb952555a/modules/imgcodecs/src/grfmt_exr.cpp#L697-L700

Comment (Kumataro) [reply modules/imgcodecs/src/loadsave.cpp]: I reconsider it. I fixed it.

It is better that this fallback do not be processed if selected encoder doesn't supports CV_8U (Maybe only OpenEXR encoder).

File: modules/imgcodecs/src/loadsave.cpp [hunk 3]
```diff
@@ -721,9 +721,9 @@ static bool imwrite_( const String& filename, const std::vector<Mat>& img_vec,
         CV_Assert( image.channels() == 1 || image.channels() == 3 || image.channels() == 4 );
 
         Mat temp;
-        if( !encoder->isFormatSupported(image.depth()) )
+        if( (!encoder->isFormatSupported(image.depth())) && (encoder->isFormatSupported(CV_8U)) )
         {
-            CV_Assert( encoder->isFormatSupported(CV_8U) );
+            CV_LOG_ONCE_WARNING(NULL, "Unsupported depth image for selected encoder is fallbacked to CV_8U.");
             image.convertTo( temp, CV_8U );
```
Comment (asmorkalov) [inline modules/imgcodecs/src/loadsave.cpp:727]: Log is good, but we get U.B., if the encoder does not support CV_8U by some reason.

**Comment by asmorkalov** [comment]
@Kumataro Thanks a lot for the contribution! I hide the function implementation in cpp file and got rid of exceptions that are thrown outside of the function. The current image i/o contract is to be exception-safe and return bool status.

File: modules/imgcodecs/src/loadsave.cpp (hunk 3 above)
Comment (Kumataro) [reply modules/imgcodecs/src/loadsave.cpp:727]: Thank you for your review. This change comes from https://github.com/opencv/opencv/pull/26211#discussion_r1779455750 .

If there are no problem, I agree with revert it.

```cpp
        if( !encoder->isFormatSupported(image.depth() ) )
        {
            CV_LOG_ONCE_WARNING(NULL, "Unsupported depth image for selected encoder is fallbacked to CV_8U.");
            CV_Assert( encoder->isFormatSupported(CV_8U) );
            image.convertTo( temp, CV_8U );
```

**Comment by sturkmen72** [comment]
@asmorkalov imwrite() writes single and multiple images. AFAIK imwritemulti() is just for python. i prefer imencode() behave same as imwrite()

File: modules/imgcodecs/include/opencv2/imgcodecs.hpp [hunk 4]
```diff
@@ -405,14 +405,28 @@ The function imencode compresses the image and stores it in the memory buffer th
 result. See cv::imwrite for the list of supported formats and flags description.
 
 @param ext File extension that defines the output format. Must include a leading period.
-@param img Image to be written.
+@param img Image to be compressed.
```
Comment (sturkmen72) [inline modules/imgcodecs/include/opencv2/imgcodecs.hpp:408]: (Mat or vector of Mat) Image or Images to be compressed.

@asmorkalov imwrite() writes single and multiple images. AFAIK imwritemulti() is just for python. also imencode() behaves same as imwrite()

Comment (sturkmen72) [reply modules/imgcodecs/include/opencv2/imgcodecs.hpp:408]: i think imwritemulti() also be defined as CV_EXPORTS_W and needs documentation

File: modules/imgcodecs/src/loadsave.cpp (hunk 3 above)
Comment (Kumataro) [reply modules/imgcodecs/src/loadsave.cpp]: Thank you for your review, I fixed it.

**Comment by Kumataro** [comment]
I'm sorry some tests for AVIF are failed. 
I'll analyze it later. Give me some time.

AVIF encoder called CV_CheckDepth() and exception is happened.
AVIF tests expects to it, but it is not able to be catched.

AVIF encoder calls CV_CheckType() with bit_depth == 6, img.depth() == CV_8U. cv::Exception will be thrown.

```cpp
    CV_CheckType(
        img.type(),
        (bit_depth == 8 && img.depth() == CV_8U) ||
            ((bit_depth == 10 || bit_depth == 12) && img.depth() == CV_16U),
        "AVIF only supports bit depth of 8 with CV_8U input or "
        "bit depth of 10 or 12 with CV_16U input");
```

AVIF tests expect that cv::Exception throws. But they doesn't.

```cpp
TEST_P(Imgcodecs_Avif_Image_EncodeDecodeSuite, imencode_imdecode) {
  const cv::Mat& img_original = get_img_original();
  ASSERT_FALSE(img_original.empty());

  // Encode.
  std::vector<unsigned char> buf;
  if (!IsBitDepthValid()) {
    EXPECT_THROW(cv::imencode(".avif", img_original, buf, encoding_params_),
                 cv::Exception);
    return;
  }
```

Maybe `loadsave.cpp` catches this exception, but it will not rethrow. So AVIF tests cannot catch.

```cpp
    try {
        if (!isMultiImg)
            code = encoder->write(write_vec[0], params);
        else
            code = encoder->writemulti(write_vec, params);

        encoder->throwOnEror();
        CV_Assert( code );
    }
    catch (const cv::Exception& e)
    {
        CV_LOG_ERROR(NULL, "imencode(): can't encode data: " << e.what());
        code = false;
    }
    catch (...)
    {
        CV_LOG_ERROR(NULL, "imencode(): can't encode data: unknown exception");
        code = false;
    }
```

I think (1) updation AVIF tests to use return value instead of exception or (2) removing try and catch in loadsave.cpp is needed.

**Comment by asmorkalov** [comment]
Yes, I changed the implementation that makes code exception safe. I propose to tune tests to handle status.

**Comment by asmorkalov** [comment]
cc @vrabaud

**Comment by Kumataro** [comment]
Thank you for your comment, and I fixed to use status code instead of catching exception.

File: modules/imgcodecs/src/grfmt_tiff.cpp [hunk 5]
```diff
@@ -1118,10 +1118,16 @@ class TiffEncoderBufHelper
...
+    static tmsize_t read( thandle_t handle, void* buffer, tmsize_t n )
     {
-        // Not used for encoding.
-        return 0;
+        // Used for imencodemulti() to stores multi-images.
+        TiffEncoderBufHelper *helper = reinterpret_cast<TiffEncoderBufHelper*>(handle);
+        size_t begin = (size_t)helper->m_buf_pos;
+        size_t end = begin + n;
+        CV_CheckGT( helper->m_buf->size(), end , "do not be over-run buffer");
+        memcpy(buffer, &(*helper->m_buf)[begin], n);
```
Comment (vrabaud) [inline modules/imgcodecs/src/grfmt_tiff.cpp:1128]: std::memcpy to stay C++

**Review by vrabaud — APPROVED** [review]

File: modules/imgcodecs/include/opencv2/imgcodecs.hpp [hunk 6]
```diff
@@ -405,14 +405,28 @@ The function imencode compresses the image and stores it in the memory buffer th
...
 @param buf Output buffer resized to fit the compressed image.
 @param params Format-specific parameters. See cv::imwrite and cv::ImwriteFlags.
 */
 CV_EXPORTS_W bool imencode( const String& ext, InputArray img,
                             CV_OUT std::vector<uchar>& buf,
                             const std::vector<int>& params = std::vector<int>());
 
+/** @brief Encodes array of images into a memory buffer.
+
+The function is analog of cv::imencode for in-memory multi-page images compression.
```
[the remark this answers is not in the document]
Comment (Kumataro) [reply modules/imgcodecs/include/opencv2/imgcodecs.hpp]: Thank you for your review, I fixed this comment.

File: modules/imgcodecs/include/opencv2/imgcodecs.hpp [hunk 7]
```diff
@@ -405,14 +405,28 @@ The function imencode compresses the image and stores it in the memory buffer th
...
+See cv::imwrite for the list of supported formats and flags description.
+
+@param ext File extension that defines the output format. Must include a leading period.
+@param imgs Vector of images be written.
```
[the remark this answers is not in the document]
Comment (Kumataro) [reply modules/imgcodecs/include/opencv2/imgcodecs.hpp]: Thank you for your review, I fixed this comment.

File: modules/imgcodecs/src/grfmt_tiff.cpp (hunk 5 above)
Comment (Kumataro) [reply modules/imgcodecs/src/grfmt_tiff.cpp]: I'm very sorry if my understanding is wrong.
I replaced from `memcpy` to `std::memcpy` in grfmt_tiff.cpp and re-indent..

File: modules/imgcodecs/test/test_exr.impl.hpp [hunk 8]
```diff
@@ -314,4 +314,27 @@ TEST(Imgcodecs_EXR, read_RGBA_unchanged)
     EXPECT_EQ(0, remove(filenameOutput.c_str()));
 }
 
+// See https://github.com/opencv/opencv/pull/26211
+// ( related with https://github.com/opencv/opencv/issues/26207 )
+TEST(Imgcodecs_EXR, imencode_regression_26207_extra)
+{
+    // CV_8U is not supported depth for EXR Encoder.
+    cv::Mat src(100, 100, CV_8UC1);
```
Comment (asmorkalov) [inline modules/imgcodecs/test/test_exr.impl.hpp:322]: I propose  to initialize mat with some values: some color or random data. It's applicable for the all new tests.
1. It makes test really deterministic
2. It makes sanitizers happy.

Comment (Kumataro) [reply modules/imgcodecs/test/test_exr.impl.hpp]: Thank you for your comment, and I agree with you !

I fixed it, and set `const` for their source images.

**Review by asmorkalov — APPROVED** [review]

**Pull request merged**